src/proto/h1/io.rs

```diff
@@ -188,7 +189,11 @@ where
         if self.read_buf_remaining_mut() < next {
             self.read_buf.reserve(next);
         }
-        let mut buf = ReadBuf::uninit(&mut self.read_buf.bytes_mut()[..]);
+        let dst = self.read_buf.bytes_mut();
+        // Safety: `dst` may include uninitialized memory, but `ReadBuf` will
+        // never uninitialize memory that has already been initialized.
+        let dst = unsafe { &mut *(dst as *mut _ as *mut [MaybeUninit<u8>]) };
+        let mut buf = ReadBuf::uninit(dst);
         match Pin::new(&mut self.io).poll_read(cx, &mut buf) {
```

Comment:
An alternative to all of this (which avoids the scary-looking pointer cast that @Urhengulas was concerned about) is to just use `tokio_util::io::poll_read_buf` from the latest `tokio-util` here:
```suggestion
        match tokio_util::io::poll_read_buf(Pin::new(&mut self.io), cx, &mut self.read_buf) {
```

That would mean adding a `tokio-util` dependency, but it's already a transitive dependency via `h2`.

Reply:
That seems like a good solution!

And this is a useful function which can't be found in the docs... (code: https://github.com/tokio-rs/tokio/blob/master/tokio-util/src/lib.rs#L67-L134).